py/builtinimport: Allow overriding of mp_builtin___import__.

laurensvalk opened the pull request:
This allows MicroPython ports without external imports enabled to override _mp_builtin___import___.

This is useful for MicroPython applications without enough resources to have `MICROPY_ENABLE_EXTERNAL_IMPORT` enabled, but which still want to allow users to do some application-specific imports beyond builtins.

Ports can already provide things like `mp_builtin_open`, so providing `mp_builtin___import__` seems to fit reasonably well.

Instead of doing it this way, it could be made more generic by changing `MICROPY_ENABLE_EXTERNAL_IMPORT`  to a configuration setting `MICROPY_ENABLE_IMPORT` with three options, such as: `MICROPY_ENABLE_IMPORT_EXTERNAL`, `MICROPY_ENABLE_IMPORT_BUILTIN_ONLY`, and  `MICROPY_ENABLE_IMPORT_CUSTOM`. 

----

Context: We are currently [overriding it in MicroPython for Pybricks](https://github.com/pybricks/micropython/commit/2c217f20f86884302fbf2dbc807e69b507b09674). I was considering adding certain new functionality (inspired by https://github.com/micropython/micropython/pull/8381), but overriding it locally without forking would be a bit cleaner and easier to maintain.

dpgeorge commented:
> This is useful for MicroPython applications without enough resources to have `MICROPY_ENABLE_EXTERNAL_IMPORT` enabled, but which still want to allow users to do some application-specific imports beyond builtins.

If a port wants to provide "custom" imports the idea is for it to enable `MICROPY_ENABLE_EXTERNAL_IMPORT` and then provide the desired functionality via a custom `mp_import_stat()`.  It doesn't need to have a filesystem.

But it looks like you actually want more sophisticated import behaviour, to support built-in packages.  #5701 intended to do this but was closed.  Maybe it needs to be revisited?  Or even the patch you link above could be merged here?

I'm not against being able to override `mp_builtin___import__()`, if it makes sense.  But I think it should be done a bit differently to match `mp_builtin_open()` in `py/builtin.h`.

laurensvalk commented:
> If a port wants to provide "custom" imports the idea is for it to enable `MICROPY_ENABLE_EXTERNAL_IMPORT` and then provide the desired functionality via a custom `mp_import_stat()`. It doesn't need to have a filesystem.

Our main resource constraint is build size. On the minimal port, `MICROPY_ENABLE_EXTERNAL_IMPORT` requires about 1372 bytes. We don't use most of it, so it is relatively expensive to add customizations that way (≈2%).

> But it looks like you actually want more sophisticated import behaviour, to support built-in packages. 

The current PR is motivated by adding another (future) customization inspired by https://github.com/micropython/micropython/pull/8381, so that's why I started looking for a more generic way to customize imports. If the adding that gets us closer in terms of build size, maybe we'll just enable external imports though.

> https://github.com/micropython/micropython/pull/5701 intended to [support built-in packages] but was closed. Maybe it needs to be revisited? Or even the patch you link above could be merged here?

The [patch](https://github.com/pybricks/micropython/commit/2c217f20f86884302fbf2dbc807e69b507b09674) we used served us well over the past years but I wasn't sure if it was generic enough to be useful in MicroPython. If you think it is, I'd be happy to make a pull request.

> I'm not against being able to override `mp_builtin___import__()`, if it makes sense. But I think it should be done a bit differently to match `mp_builtin_open()` in `py/builtin.h`.

Thanks, I'll have a look at how that works.

dpgeorge commented:
> Another option would be to add a hook that gets called just prior to raising the `ImportError`. Ports can then fill in the gap as needed.

If done this way then for it to be general there would need to be a hook at the start and end of the import function.  I think best to allow the port to override `mp_builtin___import__()`, but also for it to have access to the default one to call if it wants.

Eg have a `mp_builtin___import___default()` which is aliased to `mp_builtin___import__()` unless a board overrides the latter.

laurensvalk commented:
> I think best to allow the port to override mp_builtin___import__(), but also for it to have access to the default one to call if it wants. (...) Eg have a mp_builtin___import___default() which is aliased to mp_builtin___import__() unless a board overrides the latter.

If I'm understanding this right, this is now done with the updated version. Just let me know if you intended something else.

This approach works fine for our use case as well.

Being more general, it can now be used even if external imports are enabled (even if that isn't all that useful).

> but also for it to have access to the default one to call if it wants

One minor downside to this approach (compared to the hook approach mentioned above) is that if you just want to extend the default behavior, you have to catch the exception raised by the default handler before you can proceed.

But this is not much of a downside if external imports are disabled (which is the main use case anyway), because then `mp_builtin___import___default` is small enough that replicating it in full is not really an issue.

dpgeorge commented on py/builtin.h:
```diff
@@ -64,7 +64,15 @@ MP_DECLARE_CONST_FUN_OBJ_KW(mp_builtin_open_obj);
+#if MICROPY_MODULE_OVERRIDE_BUILTIN_IMPORT
 mp_obj_t mp_builtin___import__(size_t n_args, const mp_obj_t *args);
+#else
+#define mp_builtin___import__ mp_builtin___import___default
```
please use a static inline definition here, it's more robust than a #define

laurensvalk replied:
Done and rebased. It does increase the code size by a few bytes (+4 on minimal embedded port).

dpgeorge replied:
Ah!  So the reason is because of this line:
```c
MP_DEFINE_CONST_FUN_OBJ_VAR_BETWEEN(mp_builtin___import___obj, 1, 5, mp_builtin___import__);
```
which forces both the default and static-inline wrapper function to be linked in to the firmware (instead of just the default one).

I guess we need to go back to using the `#define` to make it efficient...

laurensvalk replied:
Thanks! I was wondering about that. I reverted it and rebased.

dpgeorge replied:
Sorry to go back and forth on this.... We don't really have a consistent way to override an arbitrary function in the core, but there is some precedence set by `py/mphal.h`.  Following the way it does things, we could do the following here:
```c
#ifndef mp_builtin___import__
#define mp_builtin___import__ mp_builtin___import___default
#endif
```
That would be pretty clean, and doesn't require an entry in `py/mpconfig.h`.  It's also how `ports/stm32/boardctrl.h` does things.

Then in your port you do:
```c
mp_obj_t custom_import(size_t n_args, const mp_obj_t *args);
#define mp_builtin___import__ custom_import
```

laurensvalk replied:
That's one of the first approaches I tried, but I got a bit stuck on where to do this exactly.

I'm able to make this work following the same `mphal` approach (the `ifndef` in `py/mphal.h` and `custom_import` in `mphalport.h`), but I figured this one doesn't belong there since it doesn't concern hardware.

It seems like the custom implementation would be provided via `mpconfigport.h`, but I'm not sure how since the declaration relies on `mp_obj_t` which isn't defined at that point.

What would be the right places to put the suggested snippets above?

dpgeorge replied:
It should go here in `py/builtin.h`.

How about this:
```c
#ifndef mp_builtin___import__
#define mp_builtin___import__ mp_builtin___import___default
#endif
mp_obj_t mp_builtin___import__(size_t n_args, const mp_obj_t *args);
mp_obj_t mp_builtin___import__default(size_t n_args, const mp_obj_t *args);
```
?

It's a bit tricky but should work.

laurensvalk replied:
Nice, that works! Updated and rebased.

dpgeorge commented:
I think the approach here is fine, it should be rare that a port wants to override this import behaviour.

laurensvalk committed:
py/builtinimport: Allow overriding of mp_builtin___import__.

This allows ports to override mp_builtin___import__.

This can be useful in MicroPython applications where
MICROPY_ENABLE_EXTERNAL_IMPORT has to be disabled due to its impact on
build size (2% to 2.5% of the minimal port). By overriding the otherwise
very minimal mp_builtin___import__, ports can still allow limited forms
of application-specific imports.

Signed-off-by: Laurens Valk <laurens@pybricks.com>

dpgeorge commented:
Thanks for updating.  Merged in d8ad87843ab38dcc74e7e86354849fff585e1ba8
